### PR TITLE
Memory Leak in Microsoft.Extensions.Caching.Memory when handling exceptions

### LINKED ISSUE #42321: Memory Leak in Microsoft.Extensions.Caching.Memory when handling exceptions
We are leaking objects when calling `MemoryCache.GetOrCreate`  and the factory method is throwing an exception.

### Repro

Run the following application, take a memory snapshot, run for a while and take another snapshot

```C#
using Microsoft.Extensions.Caching.Memory;
using System;

namespace ConsoleApp1
{
    class Program
    {
        static void Main(string[] args)
        {
            var cache = new MemoryCache(new MemoryCacheOptions());
            while (true)
            {
                try
                {
                    int value = cache.GetOrCreate("hello", entry =>
                    {
                        throw new Exception();
                        return 10;
                    });
                }
                catch
                {
                    Console.WriteLine("Get Failed");
                }
            }
        }
    }
}
```

image

### Analysis

It appears the issue is that we aren't Disposing the CacheEntry instances when an exception is being thrown:

https://github.com/dotnet/runtime/blob/33dba9518b4eb7fbc487fadc9718c408f95a826c/src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs#L98-L112

The reason we aren't calling Dispose is to fix an issue that we were caching `null` when an Exception was being thrown. See https://github.com/aspnet/Caching/pull/216.

Disposing the CacheEntry is important because every time you create a CacheEntry, it gets stuck in an AsyncLocal "stack":

https://github.com/dotnet/runtime/blob/33dba9518b4eb7fbc487fadc9718c408f95a826c/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntryHelper.cs#L28-L36

and disposing it is what "pops" it off the stack:

https://github.com/dotnet/runtime/blob/33dba9518b4eb7fbc487fadc9718c408f95a826c/src/libraries/Microsoft.Extensions.Caching.Memory/src/CacheEntryHelper.cs#L50-L63

We should always be `Disposing` the entry. That way the `ScopeLease` object is always disposed, and the stack is cleared.

However, we still need to fix the original problem: Don't cache `null` when an Exception happens. The reason this happens is because Disposing the CacheEntry is what "commits the entry into the cache". This method gets called from CacheEntry.Dispose:

https://github.com/dotnet/runtime/blob/33dba9518b4eb7fbc487fadc9718c408f95a826c/src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs#L111-L234

To fix this, we should set a flag indicating whether the `CacheEntry.Value` was ever set. If it wasn't set, we shouldn't be committing the value into the cache.

cc @Tratcher

### PR DESCRIPTION
**Pull request by eerhardt**
When an exception is thrown inside MemoryCache.GetOrCreate, we are leaking CacheEntry objects. This is because they are not being Disposed properly, and the async local CacheEntryStack is growing indefinitely.

The fix is to ensure the CacheEntry objects are disposed correctly. In order to do this, I set a flag to indicate whether the CacheEntry.Value has been set. If it hasn't, Disposing the CacheEntry won't add it to the cache.

Fix #42321

### CODE REVIEW

File: src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs [hunk 1]
```diff
@@ -43,39 +43,43 @@ public static bool TryGetValue<TItem>(this IMemoryCache cache, object key, out T
...
         {
-            ICacheEntry entry = cache.CreateEntry(key);
-            entry.AbsoluteExpiration = absoluteExpiration;
-            entry.Value = value;
-            entry.Dispose();
+            using (ICacheEntry entry = cache.CreateEntry(key))
+            {
+                entry.AbsoluteExpiration = absoluteExpiration;
+                entry.Value = value;
+            }
```
Comment (Tratcher) [inline src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs:60]: Why not the shorter synatx?
```suggestion
            using ICacheEntry entry = cache.CreateEntry(key);
            entry.AbsoluteExpiration = absoluteExpiration;
            entry.Value = value;
```

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs [hunk 2]
```diff
@@ -116,6 +116,13 @@ private void SetEntry(CacheEntry entry)
                 return;
             }
 
+            if (!entry.ValueHasBeenSet)
```
Comment (Tratcher) [inline src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:119]: Why not check this in CacheEntry.Dispose? Then you'd skip both _notifyCacheEntryDisposed and PropagateOptions if no value was was set.

File: src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs [hunk 3]
```diff
@@ -99,13 +103,11 @@ public static TItem GetOrCreate<TItem>(this IMemoryCache cache, object key, Func
         {
             if (!cache.TryGetValue(key, out object result))
             {
-                ICacheEntry entry = cache.CreateEntry(key);
-                result = factory(entry);
-                entry.SetValue(result);
-                // need to manually call dispose instead of having a using
-                // in case the factory passed in throws, in which case we
-                // do not want to add the entry to the cache
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs:107]: Does this concern no longer apply?

File: src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs [hunk 4]
```diff
@@ -99,13 +103,11 @@ public static TItem GetOrCreate<TItem>(this IMemoryCache cache, object key, Func
...
-                entry.Dispose();
+                using (ICacheEntry entry = cache.CreateEntry(key))
+                {
+                    result = factory(entry);
+                    entry.SetValue(result);
```
Comment (stephentoub) [inline src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs:109]: Why are some of these cases using `entry.Value = value;` while others are using `entry.SetValue`?  Seems like the latter is just a wrapper for the former that also returns the entry, but the return value is ignored here.

File: src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs:107]: Nevermind, it'll lack the `ValueHasBeenSet = true;` part, i.e. the purpose of the PR :) haven't had my coffee yet.

File: src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs (hunk 4 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs:109]: You are correct that's all `entry.SetValue` provides over `entry.Value`.

Fixed.

File: src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs (hunk 1 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Caching.Abstractions/src/MemoryCacheExtensions.cs:60]: Mostly memory muscle.

Fixed.

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:119]: This is a good point. I've refactored it to be this way.

Comment (wtgodbe) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs:119]: Getting some extra test failures when moving this check to `CacheEntry.Dispose`:

>   Failed   Microsoft.Extensions.Caching.Memory.CacheEntryScopeExpirationTests.LinkContextsCanNest
  Error Message:
   Assert.Null() Failure
  Expected: (null)
  Actual:   CacheEntry { AbsoluteExpiration = null, AbsoluteExpirationRelativeToNow = null, ExpirationTokens = [TestExpirationToken { ... }], Key = "myKey", PostEvictionCallbacks = [], ... }
  Stack Trace:
     at Microsoft.Extensions.Caching.Memory.CacheEntryScopeExpirationTests.LinkContextsCanNest() in /_/src/Caching/Memory/test/CacheEntryScopeExpirationTests.cs:line 341
  Failed   Microsoft.Extensions.Caching.Memory.CacheEntryScopeExpirationTests.GetWithImplicitLinkPopulatesExpirationTokens
  Error Message:
   Assert.Null() Failure
  Expected: (null)
  Actual:   CacheEntry { AbsoluteExpiration = null, AbsoluteExpirationRelativeToNow = null, ExpirationTokens = [TestExpirationToken { ... }], Key = "myKey", PostEvictionCallbacks = [], ... }
  Stack Trace:
     at Microsoft.Extensions.Caching.Memory.CacheEntryScopeExpirationTests.GetWithImplicitLinkPopulatesExpirationTokens() in /_/src/Caching/Memory/test/CacheEntryScopeExpirationTests.cs:line 317
  [xUnit.net 00:00:07.60]     Microsoft.Extensions.Caching.Memory.MemoryCacheSetAndRemoveTests.GetOrCreate_WillNotCreateEmptyValue_WhenFactoryThrows [FAIL]
  Failed   Microsoft.Extensions.Caching.Memory.MemoryCacheSetAndRemoveTests.GetOrCreate_WillNotCreateEmptyValue_WhenFactoryThrows
  Error Message:
   Assert.Null() Failure
  Expected: (null)
  Actual:   CacheEntry { AbsoluteExpiration = null, AbsoluteExpirationRelativeToNow = null, ExpirationTokens = [], Key = "myKey", PostEvictionCallbacks = [], ... }
  Stack Trace:
     at Microsoft.Extensions.Caching.Memory.MemoryCacheSetAndRemoveTests.GetOrCreate_WillNotCreateEmptyValue_WhenFactoryThrows() in /_/src/Caching/Memory/test/MemoryCacheSetAndRemoveTests.cs:line 169

**Comment by eerhardt** [comment]
I believe all current feedback has been addressed. I also renamed 2 variables to better describe what they are doing. PTAL.

File: src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs (hunk 2 above)
Comment (eerhardt) [reply src/libraries/Microsoft.Extensions.Caching.Memory/src/MemoryCache.cs]: The tests are all passing for me on my change. Did you do it the same why I did here?

**Review by Tratcher — APPROVED** [review]

**Review by JunTaoLuo — APPROVED** [review]

**Review by Tratcher — APPROVED** [review]

**Pull request merged**